packages/effect/src/internal/stream.ts

```diff
@@ -5678,6 +5678,71 @@ export const scoped = <A, E, R>(
 ): Stream.Stream<A, E, Exclude<R, Scope.Scope>> =>
   new StreamImpl(channel.ensuring(channel.scoped(pipe(effect, Effect.map(Chunk.of))), Effect.void))
 
+/** @internal */
+export const share = dual<
+  <A, E>(
+    config: Stream.ShareConfig<A, E>
+  ) => <R>(
+    self: Stream.Stream<A, E, R>
+  ) => Stream.Stream<A, E, R | Scope.Scope>,
+  <A, E, R>(
+    self: Stream.Stream<A, E, R>,
+    config: Stream.ShareConfig<A, E>
+  ) => Stream.Stream<A, E, R | Scope.Scope>
+>(
+  2,
+  <A, E, R>(self: Stream.Stream<A, E, R>, config: Stream.ShareConfig<A, E>) => {
+    let consumersCount = 0
```

Comment:
If fork the upstream inside the parent scope – there is no option to do ref counting. It will be just subscribed until the scope is closed

Reply:
As i understand, the current implementation is resource safe already. There is no resource leaks or logical mistakes.
The issue is somewhere in the forking logic. Looks like a bug in effect, i opened an issue with reproduction links in sandbox

Reply:
The current implementation is creating state eagerly in the closure. It needs to be allocated lazily / effectfully and bound to a Scope.

Reply:
I found a bug here

https://github.com/Effect-TS/effect/blob/main/packages/effect/src/internal/channel/channelExecutor.ts#L1098

Reply:
> The current implementation is creating state in the closure. It needs to be allocated lazily / effectfully and bound to a Scope.

Is it a hard requirement? I am pretty sure the state should be in the closure in this case because of a nature of this operator. And it's already bounded in the scope.
Of course i can wrap it inside `Effect.suspend`, but it will be an extra action for the user to unpack it, and i don't see benefit of it. Is there any?

Reply:
I always highly value your reviews, but in this case i would ask you to consider this implementation more thoroughly. 
Of course i will wrap it in `Effect.suspend ` if you reject, because i need this operator:) 

I'm sure that now it is implemented properly.

We don't usually need a shared stream to be shared multiple times. Because it is already shared.

It is a very rare scenario (i can't even think of a use-case) when we need to do like this

```ts
const shared = someStream.pipe(share({...}))
const shared1 = yield* shared;
const shared2 = yield* shared;
```

And even if we need to, we can do
```ts
const shared1 = someStream.pipe(share({...}))
const shared2 = someStream.pipe(share({...}))

```

But most of the times we just want to share it once and use it everywhere. In that case an extra `yield*` will be just an extra boilerplate

Reply:
It would be:

```ts
const shared = yield* someStream.pipe(share(...))

shared.pipe(...)
shared.pipe(...)
```

You run the Effect to allocate the state and resources etc